chainer/reporter.py

```diff
@@ -0,0 +1,325 @@
+                  # prediction and y.
+                  pred = self.predictor(x)
+                  abs_error = F.sum(F.abs(pred - y)) / len(x.data)
+                  loss = F.mean_squared_error(pred, y)
+
+                  # Report the mean absolute and squared error.
+                  report({'abs_error': abs_error, 'squared_error': loss}, self)
+
+                  return loss
+
+       If the link is named ``'main'`` in the hierarchy (which is the default
+       name of the target link in the
+       :class:`~chainer.training.StandardUpdater`), these reported values are
+       named ``'main/abs_error'`` and ``'main/squared_error'``. If these values
+       are reported inside the :class:`~chainer.training.extension.Evaluator`
+       extension, the names are changed to ``'validation/main/abs_error'`` and
+       ``'validation/main/squared_error'`` (``'validation'`` is the default
+       name of the Evaluator extension).
+
+    Args:
+        values (dict): Dictionary of observed values.
+        observer: Observer object. Its object ID is used to retrieve the
+            observer name, which is used as the prefix of the registration name
+            of the observed value.
+
+    """
+    if _reporters:
+        current = _reporters[-1]
+        current.report(values, observer)
+
+
+@contextlib.contextmanager
+def report_scope(observation):
+    """Returns a report scope with the current reporter.
+
+    This is equivalant to ``get_current_reporter().report_scope(observation)``,
+    except that it does not make the reporter current redundantly.
+
+    """
+    current = _reporters[-1]
+    old = current.observation
+    current.observation = observation
+    yield
+    current.observation = old
+
+
+class Summary(object):
+
+    """Online summarization of a sequence of scalars.
+
+    Summary computes the statistics of given scalars online.
+
+    """
+    def __init__(self):
+        self._x = 0
+        self._x2 = 0
+        self._n = 0
+
+    def add(self, value):
+        """Adds a scalar value.
+
+        Args:
+            value: Scalar value to accumulate. It is either a NumPy scalar or
+                a zero-dimensional array (on CPU or GPU).
+
+        """
+        with cuda.get_device(value):
+            self._x += value
+            self._x2 += value * value
+            self._n += 1
+
+    def compute_mean(self):
+        """Computes the mean."""
+        x, n = self._x, self._n
+        with cuda.get_device(x):
+            return x / n
+
+    def make_statistics(self):
+        """Computes and returns the mean and standard deviation values.
+
+        Returns:
+            tuple: Mean and standard deviation values.
+
+        """
+        x, n = self._x, self._n
+        xp = cuda.get_array_module(x)
+        with cuda.get_device(x):
+            mean = x / n
+            var = self._x2 / n - mean * mean
+            std = xp.sqrt(var, var)
```

Comment:
It makes the computation in-place (the second argument is `out`). It is valid because `var` is not used anymore.